src/conduits/RwaInputConduit3.t.sol

```diff
@@ -0,0 +1,316 @@
+        inputConduit.push(100);
+    }
+
+    function testRevertPushWIthZeroAmount() public {
+        vm.expectRevert("RwaInputConduit3/not-positive-amount");
+        inputConduit.push(0);
+    }
+
+    function testRevertPushIfInsufficientBalance() public {
+        vm.expectRevert("RwaInputConduit3/insufficient-gem-balance");
+        inputConduit.push();
+        
+        vm.expectRevert("RwaInputConduit3/insufficient-gem-balance");
+        inputConduit.push(100);
+        
+        usdx.transfer(address(inputConduit), 50 * USDX_WAD);
+        vm.expectRevert("RwaInputConduit3/insufficient-gem-balance");
+        inputConduit.push(51 * USDX_WAD);
+    }
+
+    function testPush() public {
+        assertEq(usdx.balanceOf(me), USDX_MINT_AMOUNT);
+        assertEq(usdx.balanceOf(address(inputConduit)), 0);
+        assertEq(usdx.balanceOf(address(gemA)), 0);
+
+        usdx.transfer(address(inputConduit), 500 * USDX_WAD);
+
+        assertEq(usdx.balanceOf(me), USDX_MINT_AMOUNT - 500 * USDX_WAD);
+        assertEq(usdx.balanceOf(address(inputConduit)), 500 * USDX_WAD);
+
+        assertEq(testUrn.balance(address(dai)), 0);
+
+        vm.expectEmit(true, true, false, false);
+        emit Push(address(testUrn), 500 ether);
+        inputConduit.push();
+
+        assertEq(usdx.balanceOf(address(gemA)), 500 * USDX_WAD);
+        assertEq(usdx.balanceOf(address(inputConduit)), 0);
+        assertEq(testUrn.balance(address(dai)), 500 ether);
+    }
+
+    function testPushAmount() public {
+        assertEq(usdx.balanceOf(me), USDX_MINT_AMOUNT);
+        assertEq(usdx.balanceOf(address(inputConduit)), 0);
+        assertEq(usdx.balanceOf(address(gemA)), 0);
+
+        usdx.transfer(address(inputConduit), 500 * USDX_WAD);
+
+        assertEq(usdx.balanceOf(me), USDX_MINT_AMOUNT - 500 * USDX_WAD);
+        assertEq(usdx.balanceOf(address(inputConduit)), 500 * USDX_WAD);
+
+        assertEq(testUrn.balance(address(dai)), 0);
+
+        vm.expectEmit(true, true, false, false);
+        emit Push(address(testUrn), 400 ether);
+        inputConduit.push(400 * USDX_WAD);
+
+        assertEq(usdx.balanceOf(address(gemA)), 400 * USDX_WAD);
+        assertEq(usdx.balanceOf(address(inputConduit)), 100 * USDX_WAD);
+        assertEq(testUrn.balance(address(dai)), 400 ether);
+    }
+
+    function testRevertOnSwapAboveLine() public {
+        assertEq(usdx.balanceOf(me), USDX_MINT_AMOUNT);
+        assertEq(usdx.balanceOf(address(inputConduit)), 0);
+        assertEq(usdx.balanceOf(address(gemA)), 0);
+
+        usdx.transfer(address(inputConduit), USDX_MINT_AMOUNT);
+
+        assertEq(usdx.balanceOf(me), 0);
+        assertEq(usdx.balanceOf(address(inputConduit)), USDX_MINT_AMOUNT);
+
+        assertEq(testUrn.balance(address(dai)), 0);
+
+        vm.expectRevert("Vat/ceiling-exceeded");
+        inputConduit.push();
+
+        assertEq(usdx.balanceOf(address(inputConduit)), USDX_MINT_AMOUNT);
+        assertEq(testUrn.balance(address(dai)), 0);
```

Comment:
I believe these 2 lines are not executed because `inputConduit.push()` reverts.

Reply:
Actually, it is executed, `vm.expectRevert` catch revert from the function call

Reply:
That's why we don't have Fail in test name

Reply:
Sorry, you are right.